lib/macroni/Translation/Kernel/KernelVisitor.cpp

```diff
@@ -27,69 +28,55 @@
       m_actx(actx), m_bld(bld), m_view(view) {}
 
 vast::operation kernel_visitor::visit(const vast::cg::clang_stmt *stmt,
                                       vast::cg::scope_context &scope) {
-  return visit_rcu_dereference(stmt, scope)
-      .or_else([&] { return visit_rcu_read_lock_or_unlock(stmt, scope); })
-      .or_else([&] { return visit_rcu_assign_pointer(stmt, scope); })
-      .or_else([&] { return visit_rcu_access_pointer(stmt, scope); })
-      .or_else([&] { return visit_rcu_replace_pointer(stmt, scope); })
-      // If we can't match anything, return nullptr
-      .value_or(nullptr);
-}
-
-std::optional<vast::operation>
-kernel_visitor::visit_rcu_dereference(const vast::cg::clang_stmt *stmt,
-                                      vast::cg::scope_context &scope) {
-  auto rcu_dereference = clang::dyn_cast<clang::StmtExpr>(stmt);
-  if (!rcu_dereference) {
-    return std::nullopt;
-  }
-
-  if (!m_rcu_dereference_to_p.contains(rcu_dereference) &&
-      !m_rcu_dereference_bh_to_p.contains(rcu_dereference) &&
-      !m_rcu_dereference_sched_to_p.contains(rcu_dereference)) {
-    return std::nullopt;
-  }
-
-  auto loc = m_view.location(rcu_dereference);
-  vast::cg::default_stmt_visitor visitor(m_bld, m_view, scope);
-
-  if (m_rcu_dereference_to_p.contains(rcu_dereference)) {
-    auto p = m_rcu_dereference_to_p.at(rcu_dereference);
-    auto rty = m_view.visit(p->getType(), scope);
-    auto p_op = m_view.visit(p, scope);
-    return m_bld.create<RCUDereference>(loc, rty, p_op->getOpResult(0));
-  } else if (m_rcu_dereference_bh_to_p.contains(rcu_dereference)) {
-    auto p = m_rcu_dereference_bh_to_p.at(rcu_dereference);
-    auto rty = m_view.visit(p->getType(), scope);
-    auto p_op = m_view.visit(p, scope);
-    return m_bld.create<RCUDereferenceBH>(loc, rty, p_op->getOpResult(0));
-  }
-  // m_rcu_dereference_sched_to_p.contains(rcu_dereference);
-  auto p = m_rcu_dereference_sched_to_p.at(rcu_dereference);
-  auto rty = m_view.visit(p->getType(), scope);
-  auto p_op = m_view.visit(p, scope);
-  return m_bld.create<RCUDereferenceSched>(loc, rty, p_op->getOpResult(0));
+  if (!m_expansions.contains(stmt)) {
+    return {};
+  }
+
+  auto loc = m_view.location(stmt);
+  auto expansion = m_expansions.at(stmt);
+  vast::mlir_type rty = m_bld.void_type();
+  if (auto expr = clang::dyn_cast<clang::Expr>(stmt)) {
+    rty = m_view.visit(expr->getType(), scope);
+  }
+
+  auto name = expansion.spelling.m_name;
+  if (KernelDialect::rcu_access_pointer.m_name == name ||
+      KernelDialect::rcu_dereference.m_name == name ||
+      KernelDialect::rcu_dereference_bh.m_name == name ||
+      KernelDialect::rcu_dereference_sched.m_name == name) {
+    auto p = m_view.visit(expansion.arguments[0], scope)->getResult(0);
+
+    if (KernelDialect::rcu_access_pointer.m_name == name) {
+      return m_bld.create<RCUAccessPointer>(loc, rty, p);
+    } else if (KernelDialect::rcu_dereference.m_name == name) {
```

Comment:
I would drop the `else` as the branch before always returns, but this is personal preference.

Reply:
But I would recommend it a lot here, should make the nesting a lot easier to navigate.